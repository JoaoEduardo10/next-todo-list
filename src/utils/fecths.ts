import { json } from 'stream/consumers';

export type TDataProps = {
  email: string;
  password: string;
};
interface IUser {
  name: string;
  email: string;
  password: string;
}
const urlApi = process.env.NEXT_PUBLIC_URL_API as string;

const loginUser = async (data: TDataProps) => {
  const response = await fetch(`${urlApi}/login`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  const token = await response.json();

  return token;
};

const createUser = async (user: IUser) => {
  const { email, name, password } = user;

  const response = await fetch(`${urlApi}/users`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify({ email, password, name }),
  });

  return {
    ok: response.ok,
  };
};

export { loginUser, createUser };
