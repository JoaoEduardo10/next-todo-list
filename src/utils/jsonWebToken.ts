import jwt from 'jsonwebtoken';

export interface Ijwt {
  email: string;
  name: string;
  id: string | number;
}

export interface IjwtComplete {
  iat: number;
  exp: number;
}

export const compareJwt = (token: string): (Ijwt & IjwtComplete) | null => {
  try {
    const jwtHash = process.env.NEXT_PUBLIC_HASH_JWT as string;

    const varifyToken = jwt.verify(token, jwtHash);

    return varifyToken as Ijwt & IjwtComplete;
  } catch (error: any) {
    console.log(error.message);
    return null;
  }
};
