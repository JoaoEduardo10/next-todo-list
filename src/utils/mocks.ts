import configureStore from 'redux-mock-store';
import { TBoard } from '../types';
import { mockTask } from '../components/task/mock';

const mockStore = configureStore([]);

const mockBoards: TBoard[] = [
  { id: 'board1', boardName: 'Board 1', taskConnect: '123', userId: '2' },
  { id: 'board2', boardName: 'Board 2', taskConnect: '1234', userId: '2' },
];

const store = mockStore({
  boards: {
    allBoards: [
      { id: 'board1', boardName: 'Board 1' },
      { id: 'board2', boardName: 'Board 2' },
    ],
    actualBoard: { id: 'board1', boardName: 'Board 1' },
    actualBoardWithTasks: {
      boardName: 'Board 1',
      id: '1234',
      taskConnect: '1234',
      tasks: mockTask,
      userId: '1234',
    },
  },
});

export { store, mockBoards };
