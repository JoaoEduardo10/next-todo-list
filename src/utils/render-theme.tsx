import { render, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/theme';
import { store as storeActual } from '../app/store';
import { Provider } from 'react-redux';

export const renderTheme = (
  children: React.ReactNode,
  store?: any,
): RenderResult => {
  return render(
    <Provider store={store ?? storeActual}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>,
  );
};
