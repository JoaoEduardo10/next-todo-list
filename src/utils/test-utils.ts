import { configureStore } from '@reduxjs/toolkit';
import rootReducer from '../../src/app/store';

export const createTestStore = () => {
  const store = configureStore({
    reducer: rootReducer,
    devTools: false,
  });

  return store;
};
