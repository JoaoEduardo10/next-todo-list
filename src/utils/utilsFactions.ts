import { Tasks } from '../components/task/styles';
import { TTasks } from '../types';

const returnedSubTasksConcluded = (task: TTasks) => {
  const allSubTaskConcluded = task.subTasks.filter((total) => {
    return total.concluded == true;
  });

  return allSubTaskConcluded.length;
};

const returnTasksByStatus = (tasks: TTasks[]) => {
  const tasksConcluidedFilter = tasks.filter((task) => {
    return task.status == 'concluded';
  });

  const tasksPendingFilter = tasks.filter((task) => {
    return task.status == 'pending';
  });

  const tasksProgressFilter = tasks.filter((task) => {
    return task.status == 'progress';
  });

  return { tasksConcluidedFilter, tasksPendingFilter, tasksProgressFilter };
};

export { returnedSubTasksConcluded, returnTasksByStatus };
