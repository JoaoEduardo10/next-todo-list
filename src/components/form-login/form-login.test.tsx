import {
  fireEvent,
  queryByLabelText,
  screen,
  waitFor,
} from '@testing-library/react';
import { FormLogin } from '.';
import { renderTheme } from '../../utils/render-theme';
import { act } from 'react-dom/test-utils';
import * as nextAuthReact from 'next-auth/react';

jest.useFakeTimers();
jest.mock('next-auth/react');

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: '',
      asPath: '/',
      push: jest.fn(),
    };
  },
}));

jest.spyOn(require('next/router'), 'useRouter');

const useRouter = jest.spyOn(require('next/router'), 'useRouter');

const nextAuthReactMocked = nextAuthReact as jest.Mocked<typeof nextAuthReact>;

describe('<FormLogin />', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should to render a form', () => {
    renderTheme(<FormLogin />);

    const form = screen.getByRole('form');

    expect(form).toBeInTheDocument();
  });

  it('should return an error for not typing the email', async () => {
    renderTheme(<FormLogin />);

    const form = screen.getByRole('form');

    const inputPassword = screen.getByPlaceholderText('Senha');

    fireEvent.change(inputPassword, { target: { value: '1234' } });

    fireEvent.submit(form);
    const message = await screen.findByLabelText('Message Error');

    expect(message).toHaveTextContent('Email e senha s찾o obrigat처rios!');
    expect(message).toBeInTheDocument();
  });

  it('should return an error for not typing the password', async () => {
    renderTheme(<FormLogin />);

    const form = screen.getByRole('form');

    const inputEmail = screen.getByPlaceholderText('Email');

    fireEvent.change(inputEmail, { target: { value: 'test@gmail.com' } });

    fireEvent.submit(form);
    const message = await screen.findByLabelText('Message Error');

    expect(message).toHaveTextContent('Email e senha s찾o obrigat처rios!');
    expect(message).toBeInTheDocument();
  });

  it('should make error message disappear', async () => {
    renderTheme(<FormLogin />);

    const form = screen.getByRole('form');

    const inputEmail = screen.getByPlaceholderText('Email');

    fireEvent.change(inputEmail, { target: { value: 'test@gmail.com' } });

    fireEvent.submit(form);
    const message = await screen.findByLabelText('Message Error');

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(message).toHaveStyle({
      animation: 'hidenShowMessageError 1s ease-in-out',
    });
  });

  it('should but the user does not exist', async () => {
    renderTheme(<FormLogin />);

    const form = screen.getByRole('form');

    const inputEmail = screen.getByPlaceholderText('Email');
    const inputPassowrd = screen.getByPlaceholderText('Senha');

    fireEvent.change(inputEmail, { target: { value: 'test@gmail.com' } });
    fireEvent.change(inputPassowrd, { target: { value: 'eduj1234' } });

    fireEvent.submit(form);
    const message = await screen.findByLabelText('Message Error');

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(message).toHaveTextContent('Email Ou senha incorretos!');
    expect(message).toHaveStyle({
      animation: 'hidenShowMessageError 1s ease-in-out',
    });
  });

  it('should log the user and go to the home page', async () => {
    nextAuthReactMocked.signIn.mockImplementation(() =>
      Promise.resolve({ error: '', status: 200, ok: true, url: '' }),
    );

    useRouter.mockImplementation(() => ({
      useRouter() {
        return {
          route: '/',
          pathname: '',
          query: '',
          asPath: '/',
          push: jest.fn(),
        };
      },
    }));

    await act(async () => {
      renderTheme(<FormLogin />);
    });
    const form = screen.getByRole('form');

    const inputEmail = screen.getByPlaceholderText('Email');
    const inputPassowrd = screen.getByPlaceholderText('Senha');

    fireEvent.change(inputEmail, { target: { value: 'test@gmail.com' } });
    fireEvent.change(inputPassowrd, { target: { value: 'eduj1234' } });

    fireEvent.submit(form);

    await waitFor(() => {
      const message = screen.queryByLabelText('Message Error');

      expect(message).not.toBeInTheDocument();
    });

    act(() => {
      jest.advanceTimersByTime(5000);
    });
  });

  it('should to Match Snapchot', () => {
    renderTheme(<FormLogin />);

    const form = screen.getByRole('form');

    expect(form).toMatchSnapshot();
  });
});
