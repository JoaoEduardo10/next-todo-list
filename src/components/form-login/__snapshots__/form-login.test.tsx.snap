// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormLogin /> should To match snapshot 1`] = `
.c4 {
  margin-top: 3rem;
  width: 16rem;
  padding: 1.6rem;
  background-color: #827FD3;
  border: none;
  border-radius: 1.6rem;
  font-size: 1.8rem;
  line-height: 100%;
  color: #FFFFFF;
  cursor: pointer;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
}

.c4:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c4:disabled {
  background-color: #F4F7FD;
  cursor: not-allowed;
  color: #000;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3 {
  width: 80%;
  padding: 1.6rem;
  border: none;
  background-color: #FFFFFF;
  outline: none;
  line-height: 1.2;
  font-size: 1.7rem;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1.6rem;
  border: none;
  background-color: #FFFFFF;
  gap: 0.5rem;
  box-shadow: 1rem 1rem 1.6rem #000;
  border-radius: 0.9rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 1.2;
  -webkit-transition: all 300ms ease-in;
  transition: all 300ms ease-in;
}

.c2 svg {
  font-size: 2.6rem;
  cursor: pointer;
}

.c0 {
  background-color: #F4F7FD;
  width: 40%;
  height: 90%;
  padding: 4.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 4rem;
  border-radius: 0.5rem;
  box-shadow: 0.1rem 0.1rem 0.2rem #FFFFFF;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin-bottom: 6rem;
  color: #827FD3;
  font-family: Arial,Helvetica,sans-serif;
  font-size: 3.8rem;
}

<form
  class="c0"
  role="form"
>
  <h1
    class="c1"
  >
    Login User
  </h1>
  <label
    aria-label="Label conteiner"
    class="c2"
    for="email"
  >
    <input
      class="c3"
      id="email"
      placeholder="Email"
      type="email"
      value=""
    />
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
      />
    </svg>
  </label>
  <label
    aria-label="Label conteiner"
    class="c2"
    for="password"
  >
    <input
      class="c3"
      id="password"
      placeholder="Senha"
      type="password"
      value=""
    />
    <svg
      aria-label="Password close"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath>
          <path
            clip-rule="evenodd"
            d="M124-288l388-672 388 672H124z"
            fill="none"
          />
        </clippath>
      </defs>
      <path
        d="M508 624a112 112 0 0 0 112-112c0-3.28-.15-6.53-.43-9.74L498.26 623.57c3.21.28 6.45.43 9.74.43zm370.72-458.44L836 122.88a8 8 0 0 0-11.31 0L715.37 232.23Q624.91 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 0 0 0 51.5q56.7 119.43 136.55 191.45L112.56 835a8 8 0 0 0 0 11.31L155.25 889a8 8 0 0 0 11.31 0l712.16-712.12a8 8 0 0 0 0-11.32zM332 512a176 176 0 0 1 258.88-155.28l-48.62 48.62a112.08 112.08 0 0 0-140.92 140.92l-48.62 48.62A175.09 175.09 0 0 1 332 512z"
      />
      <path
        d="M942.2 486.2Q889.4 375 816.51 304.85L672.37 449A176.08 176.08 0 0 1 445 676.37L322.74 798.63Q407.82 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 0 0 0-51.5z"
      />
    </svg>
  </label>
  <button
    class="c4"
  >
    Entrar
  </button>
</form>
`;
